tensorboard/tools/import_google_fonts.py

```diff
@@ -0,0 +1,213 @@
+
+
+def open_url(url):
+  ru = urlparse.urlparse(url)
+  pu = urlparse.ParseResult('', '', ru.path, ru.params, ru.query, ru.fragment)
+  if ru.scheme == 'https':
+    c = httplib.HTTPSConnection(ru.netloc)
+  else:
+    c = httplib.HTTPConnection(ru.netloc)
+  c.putrequest('GET', pu.geturl())
+  c.putheader('User-Agent', FLAGS.user_agent)
+  c.endheaders()
+  return c.getresponse()
+
+
+def get_sha256(fp):
+  hasher = hashlib.sha256()
+  for chunk in iter(lambda: fp.read(8 * 1024), ''):
+    hasher.update(chunk)
+  return hasher.hexdigest()
+
+
+def get_mirror_url(original):
+  return SCHEME_PATTERN.sub(FLAGS.mirror, original)
+
+
+def get_css(m):
+  url = m.group('url')
+  path = os.path.dirname(FLAGS.path) + url[url.rindex('/'):]
+  return m.group('css').replace(url, path)
+
+
+def underify(g):
+  return '_'.join(NON_REPO_PATTERN.sub('_', s.lower()) for s in g if s)
+
+
+def add_inline_file(lines, inner_lines):
+  for line in inner_lines:
+    lines.append('        %r,' % line)
+
+
+def get_html_file(css):
+  result = []
+  result.extend(FLAGS.license_html.split('\n'))
+  result.append('<style>')
+  for code in css:
+    result.extend(code.split('\n'))
+  result.append('</style>')
+  return result
+
+
+def get_extra_build_file_content(html):
+  result = [
+      'load("@io_bazel_rules_closure//closure:defs.bzl", "web_library")',
+      '',
+      'web_library(',
+      '    name = "%s",' % FLAGS.repo,
+      '    path = "%s",' % os.path.dirname(FLAGS.path),
+      '    srcs = [',
+      '        "%s",' % os.path.basename(FLAGS.path),
+      '        ":files",',
+      '    ],',
+      ')',
+      '',
+      'genrule(',
+      '    name = "html",',
+      '    outs = ["%s"],' % os.path.basename(FLAGS.path),
+      '    cmd = "\\n".join([',
+      '        "cat <<\'EOF\' >$@",',
+  ]
+  add_inline_file(result, html)
+  result.append('        "EOF",')
+  result.append('    ]),')
+  result.append(')')
+  return result
+
+
+def main(unused_argv=None):
+  assets = []
+  for url in FLAGS.urls.split(';'):
+    for m in CSS_PATTERN.finditer(open_url(url).read()):
+      assets.append(m)
+  assets.sort(key=lambda m: (m.group('family'),
+                        m.group('name'),
+                        m.group('language')))
+
+  sys.stdout.write(
+      'filegroup_external(\n'
+      '    name = "%s",\n'
+      '    licenses = ["%s"],  # %s\n'
+      '    sha256_urls = {\n' %
+      (FLAGS.repo, FLAGS.license, FLAGS.license_summary))
+
+  css = []
+  for m in assets:
+    css.append(get_css(m))
+    sys.stdout.write(
+        '        # %s (%s)\n'
+        '        "%s": [\n'
+        '            \"%s\",\n'
+        '            \"%s\",\n'
+        '        ],\n' %
+        (m.group('name'),
+         m.group('language') or 'all',
+         get_sha256(open_url(m.group('url'))),
+         get_mirror_url(m.group('url')),
+         m.group('url')))
+
+  sys.stdout.write(
+      '    },\n'
+      '    generated_rule_name = "files",\n'
```

Comment:
optional nit: Maybe make use of text substitution (ie, `%s`) instead of concatenating.

Reply:
I don't follow? There aren't any variables being inlined here.

Reply:
I mean maybe combine several of these `stdout.write`s, ie

```
output = '''
},\n'
    generated_rule_name = "files",\n
    extra_build_file_content = "\\n".join([\n
    %s
    ]),\n
)\n\n
''' % inline_file_content
```

so we can see the whole BUILD rule more easily, but upon closer inspection, that's non-trivial to do, and `add_inline_file` nicely encapsulates logic for writing content, so never mind - that's pretty organized.